src/main/java/com/spotify/sparkey/ByteBufferCleaner.java

```diff
@@ -1,62 +1,91 @@
     try {
-      Class.forName("sun.misc.Cleaner");
-      Class.forName("java.nio.DirectByteBuffer")
-              .getMethod("cleaner");
-      v = true;
-    } catch (Exception e) {
-      v = false;
+      if (System.getProperty("java.specification.version", "99").startsWith("1.")) {
+        return new ByteBufferCleaner.OldCleaner();
+      } else {
+        return new NewCleaner();
+      }
+    } catch(Exception e) {
+      throw new Error(e);
     }
-    UNMAP_SUPPORTED = v;
   }
 
-  /**
-   * Try to unmap the buffer, this method silently fails if no support
-   * for that in the JVM. On Windows, this leads to the fact,
-   * that mmapped files cannot be modified or deleted.
-   */
   public static void cleanMapping(final MappedByteBuffer buffer) {
-    if (!UNMAP_SUPPORTED) {
-      return;
-    }
+    CLEANER.clean(buffer);
+  }
 
-    try {
-      AccessController.doPrivileged(new PrivilegedExceptionAction<Object>() {
-        public Object run() throws Exception {
-          final Method getCleanerMethod = buffer.getClass()
-                  .getMethod("cleaner");
-          getCleanerMethod.setAccessible(true);
-
-          final Object cleaner = getCleanerMethod.invoke(buffer);
-          if (cleaner != null) {
-            cleaner.getClass().getMethod("clean").invoke(cleaner);
-          }
-          return null;
+  private interface Cleaner {
+    void clean(MappedByteBuffer byteBuffer);
+  }
+
+  private static class OldCleaner implements Cleaner {
+
+    @Override
+    public void clean(final MappedByteBuffer byteBuffer) {
+      try {
+        final Method getCleanerMethod  = byteBuffer.getClass().getMethod("cleaner");
+        getCleanerMethod .setAccessible(true);
+        final Object cleaner = getCleanerMethod .invoke(byteBuffer);
+        if (cleaner != null) {
+          cleaner.getClass().getMethod("clean").invoke(cleaner);
         }
-      });
-    } catch (PrivilegedActionException e) {
-      final IOException ioe = new IOException("unable to unmap the mapped buffer");
-      ioe.initCause(e.getCause());
-      e.printStackTrace(System.err);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      } catch (InvocationTargetException e) {
+        throw new RuntimeException(e);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private static class NewCleaner implements Cleaner {
+
+    private final Method clean;
+    private final Object theUnsafe;
+
+    public NewCleaner() throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException, NoSuchMethodException {
+      Class<?> unsafeClass;
+      try {
+        unsafeClass = Class.forName("sun.misc.Unsafe");
+      } catch(Exception ex) {
+        // jdk.internal.misc.Unsafe doesn't yet have an invokeCleaner() method,
+        // but that method should be added if sun.misc.Unsafe is removed.
+        unsafeClass = Class.forName("jdk.internal.misc.Unsafe");
```

Comment:
works on recent java versions?

Reply:
Travis builds this for jdk7, 8 and 11 - haven't tried anything newer.